### PR TITLE
Add NWBDataInterface

### LINKED ISSUE #348: LFP reading broken
## Bug

If you are reporting a bug please provide the following:
### Steps to Reproduce
```python
# setup
fname = 'YutaMouse41-150903'
fpath = '../../Buzsaki/SenzaiBuzsaki2017/' + fname

session_description = 'simulated MEC and LEC data'
identifier = fname
session_start_time = datetime(2015, 7, 31)
institution = 'NYU'
lab = 'Buzsaki'

source=fname

nwbfile = NWBFile(source, session_description, identifier,
                  session_start_time, datetime.now(),
                  institution=institution, lab=lab)

device = nwbfile.create_device('name', 'source')
electrode_group = nwbfile.create_electrode_group('name',
                                                'source',
                                                'description',
                                                device=device,
                                                location='location')
nwbfile.add_electrode(1,0.,0.,0.,
                      imp=0.,
                      location='.',
                      filtering = '.',
                      description = '.',
                      group=electrode_group)

region = nwbfile.create_electrode_table_region([0], '.')

elec_series = ElectricalSeries('name','source',[1,2,3],region,
                               resolution=0.,conversion=1.,
                              starting_time=0., rate=1.)

lfp = LFP('name', elec_series)
module = nwbfile.create_processing_module('0', 'source', '0')
module.add_container(lfp)

# this works
io = NWBHDF5IO('test.h5', mode='w')
io.write(nwbfile)
io.close()
```
now try to read it:
```python
io = NWBHDF5IO('test.h5')
io.read()
```

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/dev/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
    640         try:
--> 641             obj = cls(*args, **kwargs)
    642         except Exception as ex:

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
--> 345                     raise_from(TypeError(msg), None)
    346                 return func(self, **parsed['args'])

~/anaconda3/envs/dev_pynwb/lib/python3.6/site-packages/six.py in raise_from(value, from_value)

TypeError: incorrect type for 'electrical_series' (got 'list', expected 'ElectricalSeries')

The above exception was the direct cause of the following exception:

Exception                                 Traceback (most recent call last)
<ipython-input-5-e8babe8aa366> in <module>()
      1 io = NWBHDF5IO('test.h5')
----> 2 io.read()

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
    345                     raise_from(TypeError(msg), None)
--> 346                 return func(self, **parsed['args'])
    347         else:
    348             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/form/backends/io.py in read(self, **kwargs)
     30     def read(self, **kwargs):
     31         f_builder = self.read_builder()
---> 32         container = self.__manager.construct(f_builder)
     33         return container
     34 

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
    345                     raise_from(TypeError(msg), None)
--> 346                 return func(self, **parsed['args'])
    347         else:
    348             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
     66         result = self.__containers.get(builder_id)
     67         if result is None:
---> 68             result = self.__type_map.construct(builder, self)
     69             self.prebuilt(result, builder)
     70         return result

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
    345                     raise_from(TypeError(msg), None)
--> 346                 return func(self, **parsed['args'])
    347         else:
    348             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
   1015                              % str(container.__class__.__name__))  # noqa: F821
   1016         else:
-> 1017             return attr_map.construct(builder, build_manager)
   1018 
   1019     @docval({"name": "container", "type": Container, "doc": "the container to convert to a Builder"},

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
    345                     raise_from(TypeError(msg), None)
--> 346                 return func(self, **parsed['args'])
    347         else:
    348             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
    615         cls = manager.get_cls(builder)
    616         # gather all subspecs
--> 617         subspecs = self.__get_subspec_values(builder, self.spec, manager)
    618         # get the constructor argument each specification corresponds to
    619         const_args = dict()

~/dev/pynwb/src/pynwb/form/build/map.py in __get_subspec_values(self, builder, spec, manager)
    601                             ret[subspec] = val
    602                     else:
--> 603                         result = self.__get_subspec_values(sub_builder, subspec, manager)
    604                         ret.update(result)
    605         else:

~/dev/pynwb/src/pynwb/form/build/map.py in __get_subspec_values(self, builder, spec, manager)
    592                     if self.__data_type_key in sub_builder.attributes or \
    593                        not (subspec.data_type_inc is None and subspec.data_type_def is None):
--> 594                         val = manager.construct(sub_builder)
    595                         if subspec.is_many():
    596                             if subspec in ret:

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
    345                     raise_from(TypeError(msg), None)
--> 346                 return func(self, **parsed['args'])
    347         else:
    348             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
     66         result = self.__containers.get(builder_id)
     67         if result is None:
---> 68             result = self.__type_map.construct(builder, self)
     69             self.prebuilt(result, builder)
     70         return result

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
    345                     raise_from(TypeError(msg), None)
--> 346                 return func(self, **parsed['args'])
    347         else:
    348             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
   1015                              % str(container.__class__.__name__))  # noqa: F821
   1016         else:
-> 1017             return attr_map.construct(builder, build_manager)
   1018 
   1019     @docval({"name": "container", "type": Container, "doc": "the container to convert to a Builder"},

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
    345                     raise_from(TypeError(msg), None)
--> 346                 return func(self, **parsed['args'])
    347         else:
    348             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
    615         cls = manager.get_cls(builder)
    616         # gather all subspecs
--> 617         subspecs = self.__get_subspec_values(builder, self.spec, manager)
    618         # get the constructor argument each specification corresponds to
    619         const_args = dict()

~/dev/pynwb/src/pynwb/form/build/map.py in __get_subspec_values(self, builder, spec, manager)
    592                     if self.__data_type_key in sub_builder.attributes or \
    593                        not (subspec.data_type_inc is None and subspec.data_type_def is None):
--> 594                         val = manager.construct(sub_builder)
    595                         if subspec.is_many():
    596                             if subspec in ret:

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
    345                     raise_from(TypeError(msg), None)
--> 346                 return func(self, **parsed['args'])
    347         else:
    348             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
     66         result = self.__containers.get(builder_id)
     67         if result is None:
---> 68             result = self.__type_map.construct(builder, self)
     69             self.prebuilt(result, builder)
     70         return result

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
    345                     raise_from(TypeError(msg), None)
--> 346                 return func(self, **parsed['args'])
    347         else:
    348             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
   1015                              % str(container.__class__.__name__))  # noqa: F821
   1016         else:
-> 1017             return attr_map.construct(builder, build_manager)
   1018 
   1019     @docval({"name": "container", "type": Container, "doc": "the container to convert to a Builder"},

~/dev/pynwb/src/pynwb/form/utils.py in func_call(*args, **kwargs)
    344                     msg = ', '.join(parse_err)
    345                     raise_from(TypeError(msg), None)
--> 346                 return func(self, **parsed['args'])
    347         else:
    348             def func_call(*args, **kwargs):

~/dev/pynwb/src/pynwb/form/build/map.py in construct(self, **kwargs)
    643             details = '%s: %s (%s)' % (ex, cls.__name__, builder.name)
    644             msg = 'Could not construct %s object\n    %s' % (cls.__name__, details)
--> 645             raise_from(Exception(msg), ex)
    646         return obj
    647 

~/anaconda3/envs/dev_pynwb/lib/python3.6/site-packages/six.py in raise_from(value, from_value)

Exception: Could not construct LFP object
    incorrect type for 'electrical_series' (got 'list', expected 'ElectricalSeries'): LFP (LFP)
```
### Environment
Please describe your environment according to the following bullet points.
- **Python Executable:** Conda
- **Python Version:** Python 3.6
- **Operating System:** macOS
- **Pynwb Version:** 0.2.0 and dev

Any idea what's going on here? Am I setting up the data properly?

### PR DESCRIPTION
**Pull request by ajtritt**
## Motivation

The structure of /acqusition and the type requirements were too rigid. To address this, the following modifications were made:

1. Create **NWBDataInterface**, a neurodata_type that specializes **NWBContainer** to be used for representing non-metadata data objects.  Types that were formerly **Interfaces** (in NWB 1.x) and **TimeSeries** are now subtypes of **NWBDataInterface**. 
2. **NWBDataInterface** objects can be added NWBFiles as acquisition data, or as members of a **ProcessingModule** object.
3. Create a new **NWBDataInterface** called **Images** for storing a collection of images. This is not meant to replace **ImageSeries**. This interface replaces acquisition/images and provides a more general container for use elsewhere in NWB.

also fix #348 

## Checklist

- [ ] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [ ] Have you ensured the PR description clearly describes problem and the solution?
- [ ] Is your contribution compliant with our coding style ? This can be checked running `flake8` from the source directory.
- [ ] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [ ] Have you included the relevant issue number using `#XXX` notation where `XXX` is the issue number ?

### CODE REVIEW

**Comment by ajtritt** [comment]
@bendichter I would request your review, but I don't think you are a member of NeurodataWithoutBorders.

**Comment by bendichter** [comment]
So what would the workflow be for adding lfp to acquisition?

File: src/pynwb/base.py [hunk 1]
```diff
@@ -49,16 +51,32 @@ def add_container(self, **kwargs):
...
-    def get_container(self, **kwargs):
+    def get_data_interface(self, **kwargs):
         '''
         Retrieve an NWBContainer from this ProcessingModule
         '''
         container_name = getargs('container_name', kwargs)
         return self.__containers.get(container_name)
 
+    @docval({'name': 'container', 'type': NWBDataInterface, 'doc': 'the NWBDataInterface to add to this Module'})
+    def add_container(self, **kwargs):
```
Comment (oruebel) [inline src/pynwb/base.py:62]: should this function be called ``add_data_interface`` instead for consistency

**Review by oruebel — DISMISSED** [review]
Aside from the non-critical ``add_container`` vs. ``add_data_interface`` comment I added this looks good to me. I'll add and issue on the nwb_schema repo to make sure we document this change in the release notes and also update format docs as needed.

File: src/pynwb/base.py (hunk 1 above)
Comment (ajtritt) [reply src/pynwb/base.py:62]: That method exists at line 45

**Comment by ajtritt** [comment]
@bendichter I'm in the process of adding this to the documentation, but here's a snippet from docs/code/creating-and-writing-nwbfile.py:

```python
    from pynwb.ecephys import LFP
    from pynwb.behavior import Position

    lfp = f.add_acquisition(LFP('a hypothetical source'))
    ephys_ts = lfp.create_electrical_series('test_ephys_data',
                                            'an hypothetical source',
                                            ephys_data,
                                            electrode_table_region,
                                            timestamps=ephys_timestamps,
                                            # Alternatively, could specify starting_time and rate as follows
                                            # starting_time=ephys_timestamps[0],
                                            # rate=rate,
                                            resolution=0.001,
                                            comments="This data was randomly generated with numpy, using 1234 as the seed",  # noqa: E501
                                            description="Random numbers generated with numpy.random.rand")
    f.set_epoch_timeseries([ep1, ep2], ephys_ts)

    pos = f.add_acquisition(Position('a hypothetical source'))
    spatial_ts = pos.create_spatial_series('test_spatial_timeseries',
                                           'a stumbling rat',
                                           spatial_data,
                                           'origin on x,y-plane',
                                           timestamps=spatial_timestamps,
                                           resolution=0.1,
                                           comments="This data was generated with numpy, using 1234 as the seed",
                                           description="This 2D Brownian process generated with "
                                                       "np.cumsum(np.random.normal(size=(2, len(spatial_timestamps))), axis=-1).T")  # noqa: E501
    f.set_epoch_timeseries([ep1, ep2], spatial_ts)

```

File: src/pynwb/base.py (hunk 1 above)
Comment (oruebel) [reply src/pynwb/base.py:62]: Sorry for the oversight. Looks good to me.

**Review by bendichter — APPROVED** [review]
Thanks for making these changes! This really cleans up the UI and fixes some issues that I need. Looks good to me!

**Comment by ajtritt** [comment]
@neuromusic I would like to take a similar approach for addressing the issues you've raised in #306 and #146

**Pull request merged**